projects/client/RabbitMQ.Client/src/client/impl/SocketFrameHandler.cs

```diff
@@ -1,321 +1,304 @@
-                }
-            }
-
-            if (m_socket == null && endpoint.AddressFamily != AddressFamily.InterNetworkV6)
-            {
-                m_socket = ConnectUsingIPv4(endpoint, socketFactory, connectionTimeout);
-            }
-
-            Stream netstream = m_socket.GetStream();
-            netstream.ReadTimeout  = readTimeout;
-            netstream.WriteTimeout = writeTimeout;
-
-            if (endpoint.Ssl.Enabled)
-            {
-                try
-                {
-                    netstream = SslHelper.TcpUpgrade(netstream, endpoint.Ssl);
-                }
-                catch (Exception)
-                {
-                    Close();
-                    throw;
-                }
-            }
-            m_reader = new NetworkBinaryReader(new BufferedStream(netstream));
-            m_writer = new NetworkBinaryWriter(new BufferedStream(netstream));
-
-            m_writeableStateTimeout = writeTimeout;
-        }
-
-        public AmqpTcpEndpoint Endpoint { get; set; }
-
-        public EndPoint LocalEndPoint
-        {
-            get { return m_socket.Client.LocalEndPoint; }
-        }
-
-        public int LocalPort
-        {
-            get { return ((IPEndPoint)LocalEndPoint).Port; }
-        }
-
-        public EndPoint RemoteEndPoint
-        {
-            get { return m_socket.Client.RemoteEndPoint; }
-        }
-
-        public int RemotePort
-        {
-            get { return ((IPEndPoint)LocalEndPoint).Port; }
-        }
-
-        public int ReadTimeout
-        {
-            set
-            {
-                try
-                {
-                    if (m_socket.Connected)
-                    {
-                        m_socket.ReceiveTimeout = value;
-                    }
-                }
-                catch (SocketException)
-                {
-                    // means that the socket is already closed
-                }
-            }
-        }
-
-        public int WriteTimeout
-        {
-            set
-            {
-                m_writeableStateTimeout = value;
-                m_socket.Client.SendTimeout = value;
-            }
-        }
-
-        public void Close()
-        {
-            lock (_semaphore)
-            {
-                if (!_closed)
-                {
-                    try
-                    {
-                        m_socket.Close();
-                    }
-                    catch (Exception)
-                    {
-                        // ignore, we are closing anyway
-                    }
-                    finally
-                    {
-                        _closed = true;
-                    }
-                }
-            }
-        }
-
-        public Frame ReadFrame()
-        {
-            lock (m_reader)
-            {
-                return Frame.ReadFrom(m_reader);
-            }
-        }
-
-        public void SendHeader()
-        {
-            lock (m_writer)
```

Comment:
Removed this lock and flush, since we don't require flush when not using a BufferedStream and we don't require the lock on write because we will put the full frame on the stream as one unit.